File: src/main/java/io/github/willslaunwhite/linkedlist/problems/ReorderList.java
package io.github.willslaunwhite.linkedlist.problems;

import io.github.willslaunwhite.linkedlist.ListNode;

/*
 * need to refactor so there's no return
 */

public class ReorderList {
    public ListNode reorderList(ListNode head) {
        ListNode slow = head;
        ListNode fast = head.next;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode secondHalf = slow.next;
        ListNode prev = slow.next = null;
        while (secondHalf != null) {
            ListNode next = secondHalf.next;
            secondHalf.next = prev;
            prev = secondHalf;
            secondHalf = next;
        }
        ListNode firstHalf = head;
        secondHalf = prev;
        while (secondHalf != null) {
            ListNode tempFirst = firstHalf.next;
            ListNode tempSecond = secondHalf.next;
            firstHalf.next = secondHalf;
            secondHalf.next = tempFirst;
            firstHalf = tempFirst;
            secondHalf = tempSecond;
        }
        return head;
    }
}

File: python/linked_list/problems/reorder_list.py
from typing import Optional

from linked_list.list_node import ListNode

# * need to find better way to handle types than optionals


def reorder_list(head: Optional[ListNode]) -> Optional[ListNode]:
    if not head:
        return None

    slow: Optional[ListNode] = head if head else None
    fast: Optional[ListNode] = head.next if head else None
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next if fast.next else None

    second_half: Optional[ListNode] = slow.next
    slow.next = None
    prev: Optional[ListNode] = slow.next
    while second_half:
        next: Optional[ListNode] = second_half.next
        second_half.next = prev
        prev = second_half
        second_half = next

    first_half: Optional[ListNode] = head
    second_half = prev
    while second_half:
        first_temp: Optional[ListNode] = first_half.next
        second_temp: Optional[ListNode] = second_half.next
        first_half.next = second_half
        second_half.next = first_temp
        first_half = first_temp
        second_half = second_temp

    return head
